### PR TITLE
Support inline block images from jres

### PR DESCRIPTION
**Pull request by microsoftsam**
Looks a bit like this: 

screen shot 2017-11-14 at 3 56 15 pm

Authored like so: 
- Add image base64 to a jres. 
- Reference the jres in a backtick to add the icon in the block. eg: 
```
block="teleport`__icons.agent`to player"
```

### CODE REVIEW

File: pxtblocks/blocklyloader.ts [hunk 1]
```diff
@@ -3199,4 +3212,18 @@ namespace pxt.blocks {
...
     }
+
+    let iconCanvasCache: Map<string> = {};
+    function iconToFieldImage(id: string): Blockly.FieldImage {
+        let url = iconCanvasCache[id];
+        return new Blockly.FieldImage(url, 30, 30, false, '');
+    }
+
+    function initJresIcons(blockInfo: pxtc.BlocksInfo) {
+        Object.keys(blockInfo.apis.jres).forEach((jresId) => {
```
Comment (pelikhan) [inline pxtblocks/blocklyloader.ts]: Why are you not doing this lazily?

Comment (microsoftsam) [reply pxtblocks/blocklyloader.ts]: So I don't have to pass the blockInfo all the way down the stack to the initField function.

File: pxtblocks/blocklyloader.ts [hunk 2]
```diff
@@ -585,7 +585,19 @@ namespace pxt.blocks {
     }
 
     function initField(i: any, ni: number, fn: pxtc.SymbolInfo, ns: pxtc.SymbolInfo, pre: string, right?: boolean, type?: string, nsinfo?: pxtc.SymbolInfo): any {
-        if (pre)
+        if (pre && pre.indexOf('`') > -1) {
+            // parse and create icon fields for every inline icon
+            let regex = /([^`]+|(`([^`]*)`))/gi;
+            let match: RegExpExecArray;
+            while (match = regex.exec(pre)) {
```
Comment (pelikhan) [inline pxtblocks/blocklyloader.ts]: tip: replace is easier to go through matches of a regex

File: pxtblocks/blocklyloader.ts [hunk 3]
```diff
@@ -3199,4 +3212,18 @@ namespace pxt.blocks {
             }
         }
     }
+
+    let iconCanvasCache: Map<string> = {};
+    function iconToFieldImage(id: string): Blockly.FieldImage {
+        let url = iconCanvasCache[id];
```
Comment (pelikhan) [inline pxtblocks/blocklyloader.ts]: check that URL exists

File: pxtblocks/blocklyloader.ts [hunk 4]
```diff
@@ -3199,4 +3212,18 @@ namespace pxt.blocks {
             }
         }
     }
+
+    let iconCanvasCache: Map<string> = {};
```
Comment (pelikhan) [inline pxtblocks/blocklyloader.ts]: this name is quite misleading it should be jresIconCache

File: pxtblocks/blocklyloader.ts [hunk 5]
```diff
@@ -3199,4 +3212,18 @@ namespace pxt.blocks {
...
+    function iconToFieldImage(id: string): Blockly.FieldImage {
+        let url = iconCanvasCache[id];
+        return new Blockly.FieldImage(url, 30, 30, false, '');
+    }
+
+    function initJresIcons(blockInfo: pxtc.BlocksInfo) {
+        Object.keys(blockInfo.jres).forEach((jresId) => {
```
Comment (pelikhan) [inline pxtblocks/blocklyloader.ts]: should you clear iconCanvasCache to keep it from leaking icons between project reloads?

File: pxtblocks/blocklyloader.ts [hunk 6]
```diff
@@ -585,7 +585,19 @@ namespace pxt.blocks {
     }
 
     function initField(i: any, ni: number, fn: pxtc.SymbolInfo, ns: pxtc.SymbolInfo, pre: string, right?: boolean, type?: string, nsinfo?: pxtc.SymbolInfo): any {
-        if (pre)
+        if (pre && pre.indexOf('`') > -1) {
+            // parse and create icon fields for every inline icon
+            let regex = /([^`]+|(`([^`]*)`))/gi;
```
Comment (pelikhan) [inline pxtblocks/blocklyloader.ts]: what are you trying to match here? https://rise4fun.com/Rex/NY0

Comment (microsoftsam) [reply pxtblocks/blocklyloader.ts]: text without a backticks, and text within a pair of backticks

File: pxtblocks/blocklyloader.ts (hunk 2 above)
Comment (microsoftsam) [reply pxtblocks/blocklyloader.ts]: I tired that but for some reason I couldn't use it. I'll look into it again

File: pxtblocks/blocklyloader.ts (hunk 4 above)
Comment (microsoftsam) [reply pxtblocks/blocklyloader.ts]: Yeah, legacy name, that I copied over. Will change

File: pxtblocks/blocklyloader.ts (hunk 3 above)
Comment (microsoftsam) [reply pxtblocks/blocklyloader.ts]: What if it doesn't exist? We need to return a Field image.

File: pxtblocks/blocklyloader.ts (hunk 5 above)
Comment (microsoftsam) [reply pxtblocks/blocklyloader.ts]: Yes.

File: pxtblocks/blocklyloader.ts (hunk 6 above)
Comment (pelikhan) [reply pxtblocks/blocklyloader.ts]: I see that but I don't get the logic why you match text without backtick. Can we simplify this (and document it?)

Comment (microsoftsam) [reply pxtblocks/blocklyloader.ts]: Sure

Comment (pelikhan) [reply pxtblocks/blocklyloader.ts]: You use + in between the backticks

File: pxtblocks/blocklyloader.ts (hunk 3 above)
Comment (pelikhan) [reply pxtblocks/blocklyloader.ts]: We should emit an error in the console and skip that image or put a default image

**Comment by pelikhan** [comment]
Btw, we should also support this in the rendered docs.

**Comment by pelikhan** [comment]
We are leaking a "cached" compile options that did not get cleaned up properly. If we are serializing apiinfos back and forth, we should fix that instead.

**Comment by pelikhan** [comment]
@microsoftsam tests are passing.

**Pull request merged**